### PR TITLE
Gneiss init on Linux and message interface

### PR DESCRIPTION
**Pull request by jklmnn**
- first proof-of-concept for a native init on Linux (#107)
 - message interface (#109)
 - reworked initialization and finalization (#114)

### CODE REVIEW

File: src/platform/linux/init.c [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+static list_t resource_registry;
+static component_t component;
+
+static int allocate_resources(list_t *item, unsigned size, void *arg)
+{
+    resource_t *res = (resource_t *)((*item)->content);
+    return allocate_resource(res);
+}
+
+static int find_resources(list_t *item, unsigned size, void *arg)
```
Comment (senier) [inline src/platform/linux/init.c]: Shouldn't this actually be called "setup_resource"?

File: src/platform/linux/init.c [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+    }
+    return 0;
+}
+
+int main(int argc, char *argv[])
+{
+    int status;
+    pid_t chpid;
+    if(argc != 2){
+        fprintf(stderr, "Usage: %s config\n", argv[0]);
```
Comment (senier) [inline src/platform/linux/init.c]: Maybe `Usage: %s <config>` to indicate that config is not a literal?

File: src/platform/linux/init.c [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+        return 1;
+    }
+
+    component_registry = list_new();
+    resource_registry = list_new();
+    config = read_config(argv[1]);
```
Comment (senier) [inline src/platform/linux/init.c]: Can reading the config fail?

File: test/message_server/component.adb [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+   procedure Event;
+
+   package Message_Client is new Gns.Message.Reader (Unsigned_Char, Positive, C_String, 4096, Event);
+
+   type Client_Registry is array (Integer range <>) of Gns.Message.Reader_Session;
+
+   Client  : Client_Registry (1 .. 3);
+   Message : Message_Client.Message_Buffer;
+   M_Str   : String (1 .. 4097);
```
Comment (senier) [inline test/message_server/component.adb]: 4097 elements?

File: test/message_server/component.adb [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+   procedure Check_And_Print (I : Integer);
+
+   procedure Check_And_Print (I : Integer)
+   is
+      procedure Print (S1 : in out String;
+                       S2 : in out String) with
+         Import,
+         Convention => C,
+         External_Name => "print_message";
+      Label : String := Gns.Strings.Image (I) & Character'First;
```
Comment (senier) [inline test/message_server/component.adb]: Why appending `Character'First`? Is that supposed to be a null-terminator? If so, then I'd rather use Character'Val (0) for clarity.

File: src/log/client/linux/componolit-gneiss-log-client.adb [hunk 6]
```diff
@@ -34,58 +30,27 @@ is
...
    end Finalize;
 
-   procedure Print (Msg : System.Address) with
-      Import,
-      Convention    => C,
-      External_Name => "print",
-      Global        => null;
-
-   procedure Print_With_Null_Term (S : String);
+   procedure Cat (C : in out Client_Session;
```
Comment (senier) [inline src/log/client/linux/componolit-gneiss-log-client.adb]: Why `Cat`?

File: src/message/reader/linux/message_reader.c [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+static unsigned char cache[4096];
+
+void message_reader_read(list_t *item, void *buffer, long long size)
+{
+    resource_descriptor_t *res = (resource_descriptor_t *)((*item)->content);
+    TRACE("%p %p %p %lli\n", item, res, buffer, size);
+    if(mq_receive(res->fd, cache, sizeof(cache), 0) < 0){
+        perror("mq_receive");
+    }
+    memcpy(buffer, cache, size);
```
[the remark this answers is not in the document]
Comment (jklmnn) [reply src/message/reader/linux/message_reader.c]: Buffer can be smaller than the channel size (4096) which would make the read fail (or create an overflow). Data that does not fit into buffer is discarded.

File: gneiss.gpr [hunk 8]
```diff
@@ -19,26 +19,35 @@ project Gneiss is
...
                    (Source_Platform,
                     Source_Platform & "/linux",
-                    Source_Block,
-                    Source_Block & "/linux",
-                    Source_Block_Client,
-                    Source_Block_Client & "/linux",
+                    --  Source_Block,
+                    --  Source_Block & "/linux",
+                    --  Source_Block_Client,
+                    --  Source_Block_Client & "/linux",
```
[the remark this answers is not in the document]
Comment (jklmnn) [reply gneiss.gpr]: Those are implementations for Gneiss 0.1 that do not compile with the new init.

File: src/log/client/linux/componolit-gneiss-log-client.adb (hunk 6 above)
Comment (jklmnn) [reply src/log/client/linux/componolit-gneiss-log-client.adb]: The naming refers to `strcat` which concatenates two strings by appending the second string where the first one ends (by checking its null terminator). This behaviour is also employed in `Cat` which looks for the null terminator in the internal buffer and appends the string there.

File: src/message/reader/linux/message_reader.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (senier) [reply src/message/reader/linux/message_reader.c]: OK. For the updated version using SEQPACKET we should write to the application-supplied buffer directly.

File: src/log/client/linux/componolit-gneiss-log-client.adb (hunk 6 above)
Comment (senier) [reply src/log/client/linux/componolit-gneiss-log-client.adb]: I just find the name confusing as I was thinking of the command line tool `cat`. Maybe `Concatenate` could be cleared?

File: src/platform/linux/init.c (hunk 1 above)
Comment (jklmnn) [reply src/platform/linux/init.c]: `setup_resource` is already declared by an include. The rationale is that `setup_resources` is the call to setup all resources which then calls `find_resources` which finds the matching resources by checking the label and type and then sets up the matched resource with `setup_resource`.

File: test/message_server/component.adb (hunk 4 above)
Comment (jklmnn) [reply test/message_server/component.adb]: `M_Str` is passed to a C function calling `printf`. The largest possible message received from the message interface is 4096 byte so a null terminator needs to be added.

File: src/log/client/linux/componolit-gneiss-log-client.adb (hunk 6 above)
Comment (jklmnn) [reply src/log/client/linux/componolit-gneiss-log-client.adb]: Fixed by https://github.com/Componolit/gneiss/pull/117/commits/2ae856aef468ab1c297ad4f270b3f4fc2bf95d90.

File: src/platform/linux/component.h [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+#include <list.h>
+
+#define COMPONENT_RUNNING -1
+#define COMPONENT_SUCCESS 0
+#define COMPONENT_ERROR 1
+
+#define RESOURCE_READ 1
+#define RESOURCE_WRITE 2
+#define RESOURCE_READ_WRITE 3
```
[the remark this answers is not in the document]
Comment (jklmnn) [reply src/platform/linux/component.h]: Fixed by https://github.com/Componolit/gneiss/pull/117/commits/f907b9feb4b945ce1714843aa61aa1a4bb71f046.

File: src/message/componolit-gneiss-message.ads [hunk 10]
```diff
@@ -0,0 +1,39 @@
+--
+--  @summary Message interface declarations
+--  @author  Johannes Kliemann
+--  @date    2019-11-12
+--
+--  Copyright (C) 2019 Componolit GmbH
+--
+--  This file is part of ada-interface, which is distributed under the terms of the
```
[the remark this answers is not in the document]
Comment (jklmnn) [reply src/message/componolit-gneiss-message.ads]: Fixed by https://github.com/Componolit/gneiss/pull/117/commits/43f119633dd919a90cf151bd363d8e255dd1b491.

File: src/platform/linux/config.c [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+            mode = 0;
+            if(strcmp(res->name, "resource")){
+                continue;
+            }
+            resd.type = xmlGetProp(res, "type");
+            resd.name = xmlGetProp(res, "name");
+            resd.label = xmlGetProp(res, "label");
+            mode = xmlGetProp(res, "mode");
+            if(!resd.type || !resd.name || !resd.label || !mode){
+                fprintf(stderr, "ignoring invalid resource\n");
```
[the remark this answers is not in the document]
Comment (jklmnn) [reply src/platform/linux/config.c]: Fixed by https://github.com/Componolit/gneiss/pull/117/commits/17073ef973cb9fe491fb8f5048319777c139e9dc.

File: src/platform/linux/init.c (hunk 2 above)
Comment (jklmnn) [reply src/platform/linux/init.c]: Fixed by https://github.com/Componolit/gneiss/pull/117/commits/17073ef973cb9fe491fb8f5048319777c139e9dc.

File: src/platform/linux/init.c (hunk 3 above)
Comment (jklmnn) [reply src/platform/linux/init.c]: Fixed by https://github.com/Componolit/gneiss/pull/117/commits/17073ef973cb9fe491fb8f5048319777c139e9dc.

File: src/block/client/genode/componolit-gneiss-block-client.adb [hunk 12]
```diff
@@ -192,21 +192,25 @@ is
...
-      Cxx.Block.Client.Initialize (C.Instance,
-                                   Cap,
-                                   To_C_String (C_Path),
-                                   Event_Address,
-                                   Crw_Address,
-                                   Cxx.Genode.Uint64_T (Buffer_Size));
-      if Initialized (C) then
-         C.Instance.Tag := Session_Id'Pos (Tag);
+      if not Initialized (C) then
+         Cxx.Block.Client.Initialize (C.Instance,
```
[the remark this answers is not in the document]
Comment (jklmnn) [reply src/block/client/genode/componolit-gneiss-block-client.adb]: Fixed by https://github.com/Componolit/gneiss/pull/117/commits/a0a370a7ecfce7e21d2f0c3c7790776b1ef2ca96.

File: test/message_server/component.adb (hunk 5 above)
Comment (jklmnn) [reply test/message_server/component.adb]: Fixed by https://github.com/Componolit/gneiss/pull/117/commits/014bd04f977c5264d663955ec986685d2990debb.

**Pull request merged**